### ISSUE TITLE
Use of :root inside media queries

### ISSUE THREAD

**Issue by IAmAdamTaylor** [jhildenbiddle/css-vars-ponyfill#31]
Does this ponyfill support the changing of a property declared on `:root` if it is changed with a media query?

E.g.
```
:root {
  --h1-color: green;
}
@media (max-width: 1024px) {
  :root {
    --h1-color: red;
  }
}

h1 {
  color: var(--h1-color);
}
```

I'd expect to see both values transformed so that the media query is applied to the `h1` but with the new fixed value such as:

```
h1 {
  color: green;
}

@media (max-width: 1024px) {
  h1{
    color: red;
  }
}
```

Is this currently possible? I'd assume not, since ensuring the media queries are only applied to the correct elements is the tough part.

**Comment by rtynio** [comment]
Sounds great, I think this is a very big enhancement, thank you for your time @jhildenbiddle

**Comment by jhildenbiddle** [comment]
@IAmAdamTaylor --

I've created a separate card for this enhancement within the [Main Project's Backlog](https://github.com/jhildenbiddle/css-vars-ponyfill/projects/1) and closing this issue. Just FYI, as I didn't want you to think "closed" means "ignored".

**Comment by mhemesath** [comment]
Can you give a swag on the difficulty of doing this? I'm asking because I may attempt to do this myself if you think this isn't a major overhaul to enable the enhancement. 
Thanks!

**Comment by mhemesath** [comment]
Dug into this a bit.. the code is well written and relatively easy to follow given what its doing.  My initial thought is to approach this by storing variable definition scoped by media query, if present, then writing those variables out scoped by query.  I think this can be done by [extracting out this function ](https://github.com/jhildenbiddle/css-vars-ponyfill/blob/master/src/transform-css.js#L76-L102)and letting it be reused in context of media query scope.

**Comment by jhildenbiddle** [comment]
@mhemesath --

It's doable, but it is a significant amount of work that runs the risk of negatively impacting performance to the point where the ponyfill is no longer a viable option for production.

1. The parser will need to extract custom property values from media queries. Easy enough.

2. Custom properties need to be stored with their media query associations. So instead of storing a flat list of name:value pairs:
   
   ```javascript
   {
     "--my-color": "red"
   }
   ```

   Custom property data needs to be stored something like this:

   ```javascript
   {
     "default": {
       "--my-color": "red"
     },
     "only screen and (min-width: 600px)": {
       "--my-color": "blue"
     }
   }
   ```

   Also easy enough, although we're now parsing and iterating more data which will have performance implications.

3. Media queries that contain `:root`-level custom property declarations need to be duplicated and inserted adjacent to each base-level CSS rule (those not contained within a media query) that references a custom property value specified within the media query. 

   For example:
   
   ```css
   :root {
     --my-color: red;
   }
   @media only screen and (min-width: 600px) {
     :root {
       --my-color: blue;
     }
   }
   @media only screen and (min-width: 800px) {
     :root {
       --my-color: green;
     }
   }
   
   div {
     color: var(--my-color);
   }
   
   p {
     color: var(--my-color);
   }
   ```
   
   Should produce the following CSS:
   
   ```css
   div {
     color: red;
   }
   @media only screen and (min-width: 600px) {
     div {
       color: blue;
     }
   }
   @media only screen and (min-width: 800px) {
     div {
       color: green;
     }
   }
   
   p {
     color: red;
   }
   @media only screen and (min-width: 600px) {
     p {
       color: blue;
     }
   }
   @media only screen and (min-width: 800px) {
     p {
       color: green;
     }
   }
   ```

   This is a significant increase in the amount of processing that needs to take place while generating CSS. Once this change is factored in, performance becomes my primary concern.

4. In order to match the behavior of `getComputedStyle()` and `getPropertyValue()` in modern browsers, the `cssVariables` object returned as an `options.onComplete` argument should include the active CSS custom property values *across all active media queries*. This means the ponyfill will need to call `window.matchMedia` on all media queries to determine which ones are active, then merge their custom property values in the order they appeared in the CSS to determine which values are active. For example, assuming all of the following media were active:

   ```javascript
   {
     "default": {
       "--my-background": "white",
     },
     "only screen and (min-width: 600px)": {
       "--my-color": "blue",
     },
     "only screen and (orientation: portrait))": {
       "--my-color": "green"
     }
   }
   ```

   The `cssVariables` object returned as an `options.onComplete` argument would be:

   ```javascript
   {
     "--my-background": "white",
     "--my-color": "green"
   }
   ```

5. In order to match the behavior of `setComputedStyle()` in modern browsers, the custom property values passed to `options.variables` must override declarations of that custom proprty *in all media queries*. Easy enough.

6. New optimizations will need to be put in place to prevent the above changes from severely impacting ponyfill performance to the point where it is no longer a viable option for production. For example, instead of transforming all values for all media queries, the ponyfill could only transform values for active media queries. This would allow the ponyfill to generate CSS faster, which could be critical for apps leveraging `options.watch`. The downside to this approach is that the ponyfill would need to generate new CSS every time the active media queries change (e.g. window is resized, device is rotated, etc) which would introduce a restyling delay and additional jank.

7. Tests will need to be created that verify that above behavior.

There's a lot of unpack there, but for the sake of argument let's say the above changes were made. Now consider how the performance impact of those changes would affect modern Angular/React/Vue/etc. apps using the ponyfill's `options.watch` feature to handle `<style>` elements injection on route changes. Not good.

All that being said, I'd love the be proven wrong. I just want to be up front about what my concerns are and hopefully shed some light on why this feature isn't already available.

I do have a possible workaround which I will share in a separate comment (just in case I need to link directly to it in the future).

Hope this provides some insight.

**Comment by jhildenbiddle** [comment]
I wanted to offer a a solution for those needing media query and CSS custom property support while we discuss adding this functionality to the ponyfill. This is a less-than-deal brute-force solution, but for some it may be good enough.

1. Create an object that contains your app's custom properties grouped by media queries:

   ```javascript
   const mqCustomProperties = {
     "(min-width: 600px)": {
       "--my-color": "red"
     },
     "(min-width: 800px)": {
       "--my-color": "blue"
     },
     "(min-width: 1000px)": {
       "--my-color": "green"
     }
   }
   ```

1. Create a `resize` event listener that contains a [debounced](https://davidwalsh.name/javascript-debounce-function) ponyfill call with custom properties from all active media queries passed via `options.variables`:

   ```javascript
   // Simple debounce timer
   let resizeTimer;

   window.addEventListener('resize', function(evt) {
     clearTimeout(resizeTimer);
     resizeTimer = setTimeout(function() {
       const activeMediaQueryVars = {};

       // Get custom property values for all active media queries
       for (const key in mqCustomProperties) {
         if (window.matchMedia(key).matches) {
           Object.assign(activeMediaQueryVars, mqCustomProperties[key]);
         }
       }

       // Call the ponyfill
       cssVars({
         variables: activeMediaQueryVars
       });
     }, 250);
   });
   ```

1. Done!

The above (untested) code is just an example. It won't handle media queries like `@media (-webkit-min-device-pixel-ratio: 2)` or `@media print` because these aren't tied to a resize event, but a similar approach can be used to handle these as well. You'll also likely see a styling delay (depending on how complex your CSS is) while waiting for the 250ms delay before the last resize event calls the ponyfill and new CSS is generated.

Hope this helps. As I said, less than ideal, but it's an option.

**Comment by devinrhode2** [comment]
I will probably end up testing out this 2nd approach you've outlined... sounds like it would work just fine. Before reading it, I figured if worse comes to worse, in order to support IE11, I can use some react-responsive `<MediaQuery query="(max-width: 1224px)">` and render a different style tag - which of course requires dangerouslySetInnerHTML like so:
```
                <style
                    // eslint-disable-next-line react/no-danger
                    dangerouslySetInnerHTML={{__html: `
                        div.points-display { display: none !important; }
                    `}}
                />
```
Is this something you've considered? Better or worse than the 2nd approach you've suggested?

**Comment by jhildenbiddle** [comment]
Hey @devinrhode2,

I haven't considered the approach you've mentioned because my preference is for the ponyfill to work with apps as-is as instead of asking devs to make their apps work with the ponyfill. The approach I provided above doesn't match that description either, instead asking devs to modify their app by 1) extracting custom property values defined in media queries from their existing CSS and maintain it in JavaScript instead, and 2) it calling the ponyfill when media queries become active/inactive. It works, but it's a lot of extra processing that may fall below the acceptable performance line for some apps.

All that said, I do have some good news. See the following comment for details...

**Comment by jhildenbiddle** [comment]
### This feature is now planned for the upcoming 2.0 release

The next major update of the ponyfill will include quite a few optimizations that dramatically increase the performance of the ponyfill. Much of this work is currently in progress for #63. As a result of this work, processing custom properties defined in media queries should be doable without significantly impacting performance.

To be clear, the ponyfill will remain limited to processing `:root`-level custom property declarations only, although with this feature added it will soon handle `:root`-level declarations with `@media` rules.

For example:

```css
:root {
  --color: red;
}

p {
  color: var(--color);
}

@media (min-width: 500px) {
  :root {
    --color: blue;
  }
}
```

2.x output:

```css
p {
  color: red;
}

@media (min-width: 500px) {
  p {
    color: blue;
  }
}
```

No ETA unfortunately, but I will begin work on this once #63 is complete.

**Comment by jjz4689** [comment]
Does this issue also apply to queries based on the `:root` element?

In my app, I am using a `data-theme` attribute on the `html` element to signify when a `dark` theme is to be used.

For example:

```css
:root {
  --text: #000;
  --background: #fff;
}

/* example just "inverts" the colours */
:root[data-theme='dark'] {
  --text: #fff;
  --background: #000;
}
```

I noticed that nothing changes when I have the `data-theme` set to `dark`.

**Comment by jhildenbiddle** [comment]
@jjz4689 --

Unfortunately, no.

It would be possible to support advanced `:root`-level selectors, but I'm not convinced it would be worth the effort given the added complexity and potential performance implications when workarounds already exist that handle scenarios like yours (i.e. theming). For example, you can accomplish the same top-level theme switching by enabling/disabling [alternate stylesheets](https://developer.mozilla.org/en-US/docs/Web/CSS/Alternative_style_sheets) or calling the ponyfill with new theme values using [`options.variables`](https://jhildenbiddle.github.io/css-vars-ponyfill/#/?id=variables). Another option is to limit legacy browser to the default theme only in the name of progressive enhancement.

One enhancement that could be made to make life a little simpler is to support the `media` attribute on `<style>` tags. The concept is similar to enabling/disabling alternate stylesheets mentioned above, but this would allow a similar approach with `<style>` tags:

```html
<!-- This stylesheet is ON -->
<style id="theme-default">
  :root {
    --text: #000;
    --background: #fff;
  }
</style>

<!-- This stylesheet is OFF (notice media attribute) -->
<style id="theme-dark" media="none">
  :root {
    --text: #fff;
    --background: #000;
  }
</style>
```

Switching themes would just be a matter of setting the `media` attributes, then manually calling the ponyfill or using `options.watch` to auto-detect the change and call itself. For a better example of this approach, see https://inclusive-components.design/a-theme-switcher/. If this is preferable to the other solutions suggested, create a separate issue and I'll add it to the backlog.

**Comment by bozsta** [comment]
hi,

do you have any idea about ​​a possible delivery date
concerning ":root inside media queries" feature ?

**Comment by jhildenbiddle** [comment]
@bozsta --

Realistically... no. 

If you haven't already read [my previous explanation](https://github.com/jhildenbiddle/css-vars-ponyfill/issues/31#issuecomment-467227067) of the challenges involved, it's worth a read. Long story short: it's possible, but it's a lot of work that *may* end up negatively impacting performance to the point where the ponyfill is no longer viable in production. 

Performance concerns aren't the only hold up though. I also have to consider the fact that as time marches on, legacy browser support becomes less of an issue and so it becomes harder to justify time spent supporting them. If feedback proved that this was a critical issue for folks or if we were having this discussion in 2014/2015, I'd be much more likely to take on the work. Since neither of those are the case, it seems focusing on other tasks is a better use of time. That being said, I would be happy to review and assist with a PR if someone else wanted to have a go at it.

Sorry I can't provide a more optimistic outlook for the feature. Just wanted to give you (and others) an honest answer. FWIW, I've updated [my previous comment](https://github.com/jhildenbiddle/css-vars-ponyfill/issues/31#issuecomment-474106165) stating that this feature would make its way into the upcoming 2.0 release.

Closing this issue since it most likely won't be worked on in the foreseeable future. Happy to reopen in the future if things change.
